README.md
# Bucket Vector

|       Docs                       |       Crates.io                        |
|:--------------------------------:|:--------------------------------------:|
| [![docs][docs-badge]][docs-link] | [![crates][crates-badge]][crates-link] |

[docs-badge]: https://docs.rs/bucket_vec/badge.svg
[docs-link]: https://docs.rs/bucket_vec
[crates-badge]: https://img.shields.io/crates/v/bucket_vec.svg
[crates-link]: https://crates.io/crates/bucket_vec

## ⚠️ Caution

> USE WITH CAUTION

As of now this crate has not yet been battle tested
or benchmarked to an extend where the author would recommend general production
usage. Please file bugs, suggestions or enhancements to the issue tracker of [this repository](github.com/Robbepop/bucket-vec).

## Description

A vector-like data structure that organizes its elements into a set of buckets
of fixed-capacity in order to guarantee that mutations to the bucket vector
never moves elements and thus invalidates references to them.

This is comparable to a `Vec<Box<T>>` but a lot more efficient.

## Configs

The `BucketVecConfig` trait allows to customize the internal structure of your
`BucketVec`. This allows users to fine-tune their `BucketVec` for particular
use cases.

The trait mainly controls the capacity of the first bucket and the growth rate
of the capacity of new buckets.

The default `DefaultConfig` tries to balance out the different interests
between start capacity and growth rate.

## Under the Hood

The `BucketVec` is really just a vector of `Bucket` instances.
Whenever an element is pushed to the `BucketVec` the element is pushed onto
the last `Bucket` if it isn't filled, yet.
If the last `Bucket` is filled a new `Bucket` is pushed onto the `BucketVec`
with a new capacity determined by the used bucket vector configuration.

This way the `BucketVec` never moves elements around upon inserting new elements
in order to preserve references. When a normal `Vec` is modified it can potentially
invalidate references because of reallocation of the internal buffer which
might cause severe bugs if references to the internal elements are stored
outside the `Vec`. Note that normally Rust prevents such situations so the
`BucketVec` is mainly used in the area of `unsafe` Rust where a developer
actively decides that they want or need pinned references into another data
structure.

For the same reasons as stated above the `BucketVec` does not allow to remove
or swap elements.

## Example

Looking at an example `BucketVec<i32>` with the following configuration:

- `start_capacity := 1`
- `growth_rate := 2`

We have already pushed the elements `A`,.., `K` onto it.

```
[ [A], [B, C], [D, E, F, G], [H, I, J, K, _, _, _, _] ]
```

Where `_` refers to a vacant bucket entry.

Pushing another `L`,.., `O` onto the same `BucketVec` results in:

```
[ [A], [B, C], [D, E, F, G], [H, I, J, K, L, M, N, O] ]
```

So we are full on capacity for all buckets.
The next time we push another element onto the `BucketVec` it will create a new `Bucket` with a capacity of `16` since `growth_rate == 2` and our current latest bucket already has a capacity of `8`.

```
[ [A], [B, C], [D, E, F, G], [H, I, J, K, L, M, N, O], [P, 15 x _] ]
```

Where `15 x _` denotes 15 consecutive vacant entries.

## Benchmarks

`BucketVec` fullfills its role in being a replacement for situations where
a `Vec<Box<T>>` is the naive go-to solution.
The benchmark suite is still small and not super expressive but already provides
some insights in where `BucketVec` already performs pretty well and where it
could improve.

Benchmarks have been run on a Intel(R) Core(TM) i7-6700HQ CPU @ 2.60GHz.
Note that for every of the benchmark groups (`push`, `get` and `iter`) the
most efficient configuration for the `BucketVec` has been chosen.
Some benchmarks (`get`) have shown significant difference between configs.

The following is the output of a benchmark run each operating on 10k elements.

```
bucket_vec::push              time:   [43.647 us 43.861 us 44.108 us]
bucket_vec::push_get          time:   [48.872 us 49.396 us 49.834 us]
vec_box::push                 time:   [405.37 us 410.91 us 417.20 us]
vec_value::push               time:   [25.826 us 25.915 us 26.020 us]

bucket_vec::get (fast config) time:   [17.732 us 17.782 us 17.840 us]
bucket_vec::get (mid config)  time:   [243.95 us 244.75 us 245.66 us]
bucket_vec::get (slow config) time:   [341.06 us 350.02 us 361.15 us]
vec_box::get                  time:   [14.446 us 14.485 us 14.537 us]
vec_value::get                time:   [8.7939 us 8.8105 us 8.8300 us]

bucket_vec::iter              time:   [4.4195 us 4.4316 us 4.4454 us]
vec_box::iter                 time:   [9.5925 us 9.6246 us 9.6610 us]
vec_value::iter               time:   [3.5955 us 3.6043 us 3.6142 us]

bucket_vec::iter_back         time:   [3.9804 us 3.9957 us 4.0144 us]
vec_box::iter_back            time:   [9.9677 us 9.9980 us 10.033 us]
vec_value::iter_back          time:   [3.5827 us 3.5944 us 3.6080 us]

bucket_vec::iter_mut          time:   [5.0533 us 5.0710 us 5.0909 us]
vec_box::iter_mut             time:   [13.425 us 13.845 us 14.203 us]
vec_value::iter_mut           time:   [4.0172 us 4.0473 us 4.0820 us]
```

It can be seen that `BucketVec` greatly outperforms `Vec<Box<_>>` on
`push`, `iter` and `iter_back` benchmarks.
However, for some configurations `BucketVec::get` is a lot slower than
`Vec<Box<T>>::get`. The configurations used in the benchmark are:

- `fast` config: `STARTING_CAPACITY = 16; GROWTH_RATE = 1.0;`
- `mid`  config: `STARTING_CAPACITY =  4; GROWTH_RATE = 2.0;`
- `slow` config: `STARTING_CAPACITY =  5; GROWTH_RATE = 1.5`

For other benchmarked operations the difference in performance has not been as significant as for the `get` operation.

Also `BucketVec` is approximately 50% slower than the `Vec<_>` which is the
theoretical optimum that unfortunately doesn't solve the underlying problem.

## Alternatives

Before you use this data structure make sure that you are really in need of it.
The only problem it solves is that `BucketVec::{push, push_get}` guarantee that
elements stored inside the `BucketVec` are never moved around.

- This can also be achieved by `Vec<Box<T>>` although performance is generally
  worse for the majority of operations.
- Note that this is not a solution to store trait objects in a `Vec`!
- Also note that under certain curcumstances it is possible to instead of a
standard Rust `Vec<T>` and make sure that elements won't move upon a `push`
operation by using `Vec::reserve` or `Vec::reserve_exact` with appropriate
arguments for the use case.

If none of the above alternative solutions is applicable you might consider
using `BucketVec`.

## Authors & Credits

Author: Robin Freyler (github.com/Robbepop)

Special thanks to Niklas Tittjung (github.com/lugino-emeritus) who helped me a
lot with some internal formulae.

## License

Licensed under either of

 * Apache license, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.

### Dual licence: [![badge][license-mit-badge]](LICENSE-MIT) [![badge][license-apache-badge]](LICENSE-APACHE)

[license-mit-badge]: https://img.shields.io/badge/license-MIT-blue.svg
[license-apache-badge]: https://img.shields.io/badge/license-APACHE-orange.svg


RELEASES.md
# Releases

Release history for `bucket_vec` crate.

## 0.8.0 - 2020-02-24

- Add new methods
    - `BucketVec::first` and `BucketVec::first_mut`
    - `BucketVec::last` and `BucketVec::last_mut`

## 0.7.1 - 2020-02-24

- Greatly improve performance of `BucketVec::push_get`

## 0.7 - 2020-02-23

- Implement `IntoIterator` for `&BucketVec` and `&mut BucketVec`
- Implement `parity-scale-codec::{Encode, Decode}` for `BucketVec`
    - Opt-in crate feature: `scale-1`

## 0.6 - 2020-02-23

- Losen trait bound constraints for
    - `Clone`
    - `PartialEq`
    - `Eq`
    - `PartialOrd`
    - `Ord`
    - `Hash`

    trait implementations of `BucketVec`.
- Implement `IntoIterator` for `BucketVec`

## 0.5 - 2020-02-23

- Add derive for `Clone` for `BucketVec`
- Add `BucketVec::iter_mut`
- Improve performance of `BucketVec::iter`
- Restructured crate slightly

## 0.4 - 2020-02-22

- Slightly improve performance of `BucketVec::{push, get}` for natural `GROWTH_RATE` config
- Greatly improve performance of `BucketVec::iter`
- Fix bug with `no_std` compat

## 0.3 - 2020-02-22

- Optimize `BucketVec::get` for `GROWTH_RATE` of `2.0`
- Lift trait bound constraints for `C` generic parameter for `BucketVec::iter`

## 0.2 - 2020-02-21

- derive `PartialEq`, `Eq`, `PartialOrd`, `Ord` and `Hash` for `BucketVec`
- implement `core::iter::Extend` for `BucketVec`
- export `Iter` publicly
- implement `FromIterator<Item = T>` for `BucketVec<T>`

## 0.1.1

- fix bug with `BucketVec::{get, get_mut}` index calculations
- implement `no_std` compat

## 0.1

- initial release


src/tests.rs
use super::*;

/// A configuration for bucket vectors that grows quadratically.
#[derive(Debug)]
pub enum QuadraticConfig {}

impl BucketVecConfig for QuadraticConfig {
    /// The first bucket has a capacity of 1.
    const STARTING_CAPACITY: usize = 1;
    /// The next bucket always doubles in capacity.
    const GROWTH_RATE: f64 = 3.0;
}

/// A configuration for bucket vectors that grows cubically.
#[derive(Debug)]
pub enum CubicConfig {}

impl BucketVecConfig for CubicConfig {
    /// The first bucket has a capacity of 1.
    const STARTING_CAPACITY: usize = 1;
    /// The next bucket always triples in capacity.
    const GROWTH_RATE: f64 = 3.0;
}

/// A configuration for bucket vectors that has equal bucket capacities.
#[derive(Debug)]
pub enum EqualSizeConfig {}

impl BucketVecConfig for EqualSizeConfig {
    /// The first bucket has a capacity of 4.
    const STARTING_CAPACITY: usize = 4;
    /// All buckets have the same capacity as the first bucket.
    const GROWTH_RATE: f64 = 1.0;
}

/// A configuration for bucket vectors where every bucket has a capacity of 1.
///
/// # Note
///
/// This is more or less interesting since it has similar layout or even
/// performance characteristics as if a `Vec<Box<T>>` was used instead of
/// the bucket vector.
#[derive(Debug)]
pub enum WastefulConfig {}

impl BucketVecConfig for WastefulConfig {
    /// The first bucket has a capacity of 1.
    const STARTING_CAPACITY: usize = 1;
    /// All buckets have the same capacity as the first bucket.
    const GROWTH_RATE: f64 = 1.0;
}

/// A config for bucket vectors that tries to balance interests.
#[derive(Debug)]
pub enum C3G1x5Config {}

impl BucketVecConfig for C3G1x5Config {
    /// The first bucket has a capacity of 3.
    const STARTING_CAPACITY: usize = 3;
    /// The next bucket is always approx 50% larger.
    const GROWTH_RATE: f64 = 1.5;
}

/// A crazy PI config for bucket vectors to drive to limits.
#[derive(Debug)]
pub enum CrazyPiConfig {}

impl BucketVecConfig for CrazyPiConfig {
    /// The first bucket has a capacity of approximately PI.
    const STARTING_CAPACITY: usize = 3;
    /// The next bucket is always PI larger.
    const GROWTH_RATE: f64 = 3.14159265;
}

macro_rules! create_test_for_configs {
    ( $test_fn:ident ) => {
        paste::item! {
            #[test]
            fn [<$test_fn _quadratic_config_small>]() {
                $test_fn::<QuadraticConfig>(small_test_values())
            }

            #[test]
            fn [<$test_fn _quadratic_config_big>]() {
                $test_fn::<QuadraticConfig>(big_test_values())
            }

            #[test]
            fn [<$test_fn _cubic_config_small>]() {
                $test_fn::<CubicConfig>(small_test_values())
            }

            #[test]
            fn [<$test_fn _cubic_config_big>]() {
                $test_fn::<CubicConfig>(big_test_values())
            }

            #[test]
            fn [<$test_fn _equal_size_config_small>]() {
                $test_fn::<EqualSizeConfig>(small_test_values())
            }

            #[test]
            fn [<$test_fn _equal_size_config_big>]() {
                $test_fn::<EqualSizeConfig>(big_test_values())
            }

            #[test]
            fn [<$test_fn _wasteful_config_small>]() {
                $test_fn::<WastefulConfig>(small_test_values())
            }

            #[test]
            fn [<$test_fn _wasteful_config_big>]() {
                $test_fn::<WastefulConfig>(big_test_values())
            }

            #[test]
            fn [<$test_fn _c3g1x5_config_small>]() {
                $test_fn::<C3G1x5Config>(small_test_values())
            }

            #[test]
            fn [<$test_fn _c3g1x5_config_big>]() {
                $test_fn::<C3G1x5Config>(big_test_values())
            }

            #[test]
            fn [<$test_fn _crazy_pi_config_small>]() {
                $test_fn::<CrazyPiConfig>(small_test_values())
            }

            #[test]
            fn [<$test_fn _crazy_pi_config_big>]() {
                $test_fn::<CrazyPiConfig>(big_test_values())
            }
        }
    };
}

fn small_test_values() -> Vec<i32> {
    vec![5, 42, 1337, -1, 0, 7, 66, 12, 1, 2, 3, 1]
}

fn big_test_values() -> Vec<i32> {
    let mut vec = Vec::new();
    let mut rng = rand::thread_rng();
    use rand::Rng as _;
    for _ in 0..10_000 {
        vec.push(rng.gen());
    }
    vec
}

fn new_works_for<C>(_test_values: Vec<i32>)
where
    C: BucketVecConfig,
{
    let mut vec = <BucketVec<i32, C>>::new();
    assert_eq!(vec.len(), 0);
    assert!(vec.is_empty());
    assert!(vec.iter().next().is_none());
    assert!(vec.iter().next_back().is_none());
    assert!(vec.iter_mut().next().is_none());
    assert!(vec.iter_mut().next_back().is_none());
    assert!(vec.first().is_none());
    assert!(vec.first_mut().is_none());
    assert!(vec.last().is_none());
    assert!(vec.last_mut().is_none());
}
create_test_for_configs!(new_works_for);

fn push_works_for<C>(test_values: Vec<i32>)
where
    C: BucketVecConfig,
{
    let mut vec = <BucketVec<i32, C>>::new();
    let expected_len = test_values.len();
    for (i, value) in test_values.into_iter().enumerate() {
        assert_eq!(vec.len(), i);
        vec.push(value);
    }
    assert_eq!(vec.len(), expected_len);
}
create_test_for_configs!(push_works_for);

/// Asserts that the two given iterators yield the same elements.
fn assert_iter_eq<I1, I2, T>(iter1: I1, iter2: I2)
where
    I1: Iterator<Item = T> + ExactSizeIterator,
    I2: Iterator<Item = T> + ExactSizeIterator,
    T: core::fmt::Debug + PartialEq,
{
    assert_eq!(iter1.len(), iter2.len());
    for (elem1, elem2) in iter1.zip(iter2) {
        assert_eq!(elem1, elem2);
    }
}

fn iter_next_works_for<C>(test_values: Vec<i32>)
where
    C: BucketVecConfig,
{
    let vec = test_values.iter().cloned().collect::<BucketVec<_>>();
    assert_iter_eq(vec.iter(), test_values.iter());
}
create_test_for_configs!(iter_next_works_for);

fn iter_next_back_works<C>(test_values: Vec<i32>)
where
    C: BucketVecConfig,
{
    let vec = test_values.iter().cloned().collect::<BucketVec<_>>();
    assert_iter_eq(vec.iter().rev(), test_values.iter().rev());
}
create_test_for_configs!(iter_next_back_works);

fn iter_next_meet_middle_works_for<C>(test_values: Vec<i32>)
where
    C: BucketVecConfig,
{
    let vec = test_values.iter().cloned().collect::<BucketVec<_>>();
    let mut expected = test_values.iter();
    let mut iter = vec.iter();
    // Lock-step get `next` and `next_back` from iterators.
    for step in 0..iter.len() {
        if step % 2 == 0 {
            // For every even step get `next`:
            assert_eq!(iter.next(), expected.next());
        } else {
            // For every odd step get `next_back`:
            assert_eq!(iter.next_back(), expected.next_back());
        }
    }
    // At the end `iter` should be empty:
    assert_eq!(iter.next(), None);
    assert_eq!(iter.next_back(), None);
}
create_test_for_configs!(iter_next_meet_middle_works_for);

fn iter_mut_next_works_for<C>(mut test_values: Vec<i32>)
where
    C: BucketVecConfig,
{
    let mut vec = test_values.iter().cloned().collect::<BucketVec<_>>();
    assert_iter_eq(vec.iter_mut(), test_values.iter_mut());
}
create_test_for_configs!(iter_mut_next_works_for);

fn iter_mut_next_back_works_for<C>(mut test_values: Vec<i32>)
where
    C: BucketVecConfig,
{
    let mut vec = test_values.iter().cloned().collect::<BucketVec<_>>();
    assert_iter_eq(vec.iter_mut().rev(), test_values.iter_mut().rev());
}
create_test_for_configs!(iter_mut_next_back_works_for);

fn iter_mut_next_meet_middle_works_for<C>(mut test_values: Vec<i32>)
where
    C: BucketVecConfig,
{
    let mut vec = test_values.iter().cloned().collect::<BucketVec<_>>();
    let mut expected = test_values.iter_mut();
    let mut iter = vec.iter_mut();
    // Lock-step get `next` and `next_back` from iterators.
    for step in 0..iter.len() {
        if step % 2 == 0 {
            // For every even step get `next`:
            assert_eq!(iter.next(), expected.next());
        } else {
            // For every odd step get `next_back`:
            assert_eq!(iter.next_back(), expected.next_back());
        }
    }
    // At the end `iter` should be empty:
    assert_eq!(iter.next(), None);
    assert_eq!(iter.next_back(), None);
}
create_test_for_configs!(iter_mut_next_meet_middle_works_for);

fn access_works_for<C>(test_values: Vec<i32>)
where
    C: BucketVecConfig,
{
    let mut vec = <BucketVec<i32, C>>::new();
    for (n, expected) in test_values.into_iter().enumerate() {
        assert_eq!(vec.push_get(expected).index(), 3 * n);
        assert_eq!(vec.push_get(expected).into_ref(), &expected);
        assert_eq!(vec.push_get(expected).into_mut(), &mut expected.clone());
    }
}
create_test_for_configs!(access_works_for);

fn get_works_for<C>(test_values: Vec<i32>)
where
    C: BucketVecConfig,
{
    let vec = test_values.iter().cloned().collect::<BucketVec<_>>();
    for (n, expected) in test_values.into_iter().enumerate() {
        assert_eq!(vec.get(n), Some(&expected));
    }
    assert_eq!(vec.get(vec.len()), None);
}
create_test_for_configs!(get_works_for);

fn get_mut_works_for<C>(test_values: Vec<i32>)
where
    C: BucketVecConfig,
{
    let mut vec = test_values.iter().cloned().collect::<BucketVec<_>>();
    for (n, mut expected) in test_values.into_iter().enumerate() {
        assert_eq!(vec.get_mut(n), Some(&mut expected));
    }
    assert_eq!(vec.get_mut(vec.len()), None);
}
create_test_for_configs!(get_mut_works_for);

fn last_works_for<C>(mut test_values: Vec<i32>)
where
    C: BucketVecConfig,
{
    let mut vec = test_values.iter().cloned().collect::<BucketVec<_>>();
    assert_eq!(vec.last(), test_values.last());
    assert_eq!(vec.last_mut(), test_values.last_mut());
}
create_test_for_configs!(last_works_for);

fn first_works_for<C>(mut test_values: Vec<i32>)
where
    C: BucketVecConfig,
{
    let mut vec = test_values.iter().cloned().collect::<BucketVec<_>>();
    assert_eq!(vec.first(), test_values.first());
    assert_eq!(vec.first_mut(), test_values.first_mut());
}
create_test_for_configs!(first_works_for);


src/math.rs
//! Utility module to help with `no_std` Rust not supporting certain float operations.

/// Trait to either forward implementation to `std` library or `libm` for `no_std`.
pub trait FloatExt {
    /// Calculates the absolute value for the float.
    fn abs(self) -> Self;
    /// Returns the fraction part of the float.
    fn fract(self) -> Self;
    /// Rounds the float down to the next natural representable float.
    fn floor(self) -> Self;
    /// Rounds teh float up to the next natural representable float.
    fn ceil(self) -> Self;
    /// Calculates the natural logarithm of the float to the base.
    fn log(self, base: Self) -> Self;
    /// Caululates the logarithm of the float to base 2.
    fn log2(self) -> Self;
    /// Calculates the power of the float to the integer exponent.
    fn powi(self, exp: i32) -> Self;
}

#[cfg(feature = "std")]
impl FloatExt for f64 {
    fn abs(self) -> Self {
        f64::abs(self)
    }

    fn fract(self) -> Self {
        f64::fract(self)
    }

    fn floor(self) -> Self {
        f64::floor(self)
    }

    fn ceil(self) -> Self {
        f64::ceil(self)
    }

    fn log(self, base: Self) -> Self {
        f64::log(self, base)
    }

    fn log2(self) -> Self {
        f64::log2(self)
    }

    fn powi(self, exp: i32) -> Self {
        f64::powi(self, exp)
    }
}

#[cfg(not(feature = "std"))]
impl FloatExt for f64 {
    fn abs(self) -> Self {
        libm::fabs(self)
    }

    fn fract(self) -> Self {
        self - libm::trunc(self)
    }

    fn floor(self) -> Self {
        libm::floor(self)
    }

    fn ceil(self) -> Self {
        libm::ceil(self)
    }

    fn log(self, base: Self) -> Self {
        libm::log(self) / libm::log(base)
    }

    fn log2(self) -> Self {
        libm::log2(self)
    }

    fn powi(self, exp: i32) -> Self {
        libm::pow(self, exp as f64)
    }
}


src/bucket.rs
#[cfg(not(feature = "std"))]
use alloc::vec::Vec;

#[cfg(feature = "std")]
use std::vec;

#[cfg(not(feature = "std"))]
use alloc::vec;

/// An fixed capacity bucket within the bucket vector.
#[derive(Debug, Clone)]
pub struct Bucket<T> {
    /// The entries of this bucket.
    entries: Vec<T>,
}

impl<T> Bucket<T> {
    /// Creates a new emtpy bucket with a fixed capacity.
    ///
    /// # Note
    ///
    /// This does not allocate any heap memory.
    pub fn new(capacity: usize) -> Self {
        Self {
            entries: Vec::with_capacity(capacity),
        }
    }

    /// Returns the current length of the entry.
    pub fn len(&self) -> usize {
        self.entries.len()
    }

    /// Returns the fixed capacity of the entry.
    pub fn capacity(&self) -> usize {
        self.entries.capacity()
    }

    /// Returns `true` if the entry is empty.
    #[allow(dead_code)]
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    /// Returns a shared reference to the element at the given index.
    ///
    /// # Panics
    ///
    /// Panics if the index is out of bounds.
    pub fn get(&self, index: usize) -> Option<&T> {
        self.entries.get(index)
    }

    /// Returns an exclusive reference to the element at the given index.
    ///
    /// # Panics
    ///
    /// Panics if the index is out of bounds.
    pub fn get_mut(&mut self, index: usize) -> Option<&mut T> {
        self.entries.get_mut(index)
    }

    /// Pushes a new value into the fixed capacity entry.
    ///
    /// # Panics
    ///
    /// If the entry is already at its capacity.
    /// Note that this panic should never happen since the entry is only ever
    /// accessed by its outer bucket vector that checks before pushing.
    pub fn push(&mut self, new_value: T) {
        if self.len() == self.capacity() {
            panic!("entry is already filled to capacity")
        }
        self.entries.push(new_value);
    }

    /// Returns an iterator over the entries of the bucket.
    pub fn iter(&self) -> core::slice::Iter<T> {
        self.entries.iter()
    }

    /// Returns an iterator over the entries of the bucket.
    pub fn iter_mut(&mut self) -> core::slice::IterMut<T> {
        self.entries.iter_mut()
    }
}

impl<T> IntoIterator for Bucket<T> {
    type Item = T;
    type IntoIter = vec::IntoIter<T>;

    fn into_iter(self) -> Self::IntoIter {
        self.entries.into_iter()
    }
}

impl<T> core::ops::Index<usize> for Bucket<T> {
    type Output = T;

    fn index(&self, index: usize) -> &Self::Output {
        self.get(index).expect("index out of bounds")
    }
}

impl<T> core::ops::IndexMut<usize> for Bucket<T> {
    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
        self.get_mut(index).expect("index out of bounds")
    }
}


benches/bench.rs
use bucket_vec::BucketVec;
use criterion::{black_box, criterion_group, criterion_main, BatchSize, BenchmarkId, Criterion};

/// A configuration for bucket vectors that has equal bucket capacities.
#[derive(Debug)]
pub enum EqualSizeConfig {}

impl bucket_vec::BucketVecConfig for EqualSizeConfig {
    const STARTING_CAPACITY: usize = 16;
    const GROWTH_RATE: f64 = 1.0;
}

/// A configuration for bucket vectors that tries to balance out interests.
#[derive(Debug)]
pub enum C5g1x5Config {}

impl bucket_vec::BucketVecConfig for C5g1x5Config {
    const STARTING_CAPACITY: usize = 5;
    const GROWTH_RATE: f64 = 1.5;
}

/// A configuration for bucket vectors that grows quadratically.
#[derive(Debug)]
pub enum QuadraticConfig {}

impl bucket_vec::BucketVecConfig for QuadraticConfig {
    const STARTING_CAPACITY: usize = 4;
    const GROWTH_RATE: f64 = 2.0;
}

const BIG_SAMPLE_SIZE: usize = 10_000;

fn bench_bucket_vec_push(c: &mut Criterion) {
    c.bench_with_input(
        BenchmarkId::new("bucket_vec::push", BIG_SAMPLE_SIZE),
        &BIG_SAMPLE_SIZE,
        |b, &size| {
            let mut vec = BucketVec::<i32, QuadraticConfig>::new();
            b.iter(|| {
                for _ in 0..size {
                    vec.push(0);
                }
            });
        },
    );
}

fn bench_vec_box_push(c: &mut Criterion) {
    c.bench_with_input(
        BenchmarkId::new("vec_box::push", BIG_SAMPLE_SIZE),
        &BIG_SAMPLE_SIZE,
        |b, &size| {
            let mut vec = Vec::<Box<i32>>::new();
            b.iter(|| {
                for _ in 0..size {
                    vec.push(Box::new(0));
                }
            });
        },
    );
}

fn bench_vec_value_push(c: &mut Criterion) {
    c.bench_with_input(
        BenchmarkId::new("vec_value::push", BIG_SAMPLE_SIZE),
        &BIG_SAMPLE_SIZE,
        |b, &size| {
            let mut vec = Vec::<i32>::new();
            b.iter(|| {
                for _ in 0..size {
                    vec.push(0);
                }
            });
        },
    );
}

fn bench_bucket_vec_push_get(c: &mut Criterion) {
    c.bench_with_input(
        BenchmarkId::new("bucket_vec::push_get", BIG_SAMPLE_SIZE),
        &BIG_SAMPLE_SIZE,
        |b, &size| {
            let mut vec = BucketVec::<i32, QuadraticConfig>::new();
            b.iter(|| {
                for _ in 0..size {
                    black_box(vec.push_get(0));
                }
            });
        },
    );
}

fn bench_bucket_vec_get_fast_config(c: &mut Criterion) {
    let vec = (0..BIG_SAMPLE_SIZE)
        .into_iter()
        .map(|value| value as i32)
        .collect::<BucketVec<i32, EqualSizeConfig>>();
    c.bench_with_input(
        BenchmarkId::new("bucket_vec::get (fast config)", BIG_SAMPLE_SIZE),
        &vec,
        |b, vec| {
            b.iter(|| {
                for i in 0..vec.len() {
                    black_box(vec.get(i).map(|val| *val));
                }
            });
        },
    );
}

fn bench_bucket_vec_get_medium_config(c: &mut Criterion) {
    let vec = (0..BIG_SAMPLE_SIZE)
        .into_iter()
        .map(|value| value as i32)
        .collect::<BucketVec<i32, QuadraticConfig>>();
    c.bench_with_input(
        BenchmarkId::new("bucket_vec::get (medium config)", BIG_SAMPLE_SIZE),
        &vec,
        |b, vec| {
            b.iter(|| {
                for i in 0..vec.len() {
                    black_box(vec.get(i).map(|val| *val));
                }
            });
        },
    );
}

fn bench_bucket_vec_get_slow_config(c: &mut Criterion) {
    let vec = (0..BIG_SAMPLE_SIZE)
        .into_iter()
        .map(|value| value as i32)
        .collect::<BucketVec<i32, C5g1x5Config>>();
    c.bench_with_input(
        BenchmarkId::new("bucket_vec::get (slow config)", BIG_SAMPLE_SIZE),
        &vec,
        |b, vec| {
            b.iter(|| {
                for i in 0..vec.len() {
                    black_box(vec.get(i).map(|val| *val));
                }
            });
        },
    );
}

fn bench_vec_box_get(c: &mut Criterion) {
    let vec = (0..BIG_SAMPLE_SIZE)
        .into_iter()
        .map(|value| Box::new(value as i32))
        .collect::<Vec<Box<i32>>>();
    c.bench_with_input(
        BenchmarkId::new("vec_box::get", BIG_SAMPLE_SIZE),
        &vec,
        |b, vec| {
            b.iter(|| {
                for i in 0..vec.len() {
                    black_box(vec.get(i).map(|val| **val));
                }
            });
        },
    );
}

fn bench_vec_value_get(c: &mut Criterion) {
    let vec = vec![0; BIG_SAMPLE_SIZE];
    c.bench_with_input(
        BenchmarkId::new("vec_value::get", BIG_SAMPLE_SIZE),
        &vec,
        |b, vec| {
            b.iter(|| {
                for i in 0..vec.len() {
                    black_box(vec.get(i).map(|val| *val));
                }
            });
        },
    );
}

fn bench_bucket_vec_iter(c: &mut Criterion) {
    let vec = (0..BIG_SAMPLE_SIZE)
        .into_iter()
        .map(|value| value as i32)
        .collect::<BucketVec<i32, QuadraticConfig>>();
    c.bench_with_input(
        BenchmarkId::new("bucket_vec::iter", BIG_SAMPLE_SIZE),
        &vec,
        |b, vec| {
            b.iter(|| {
                for i in vec.iter() {
                    black_box(*i);
                }
            });
        },
    );
}

fn bench_vec_box_iter(c: &mut Criterion) {
    let vec = (0..BIG_SAMPLE_SIZE)
        .into_iter()
        .map(|value| Box::new(value as i32))
        .collect::<Vec<Box<i32>>>();
    c.bench_with_input(
        BenchmarkId::new("vec_box::iter", BIG_SAMPLE_SIZE),
        &vec,
        |b, vec| {
            b.iter(|| {
                for i in vec.iter() {
                    black_box(**i);
                }
            });
        },
    );
}

fn bench_vec_value_iter(c: &mut Criterion) {
    let vec = vec![0; BIG_SAMPLE_SIZE];
    c.bench_with_input(
        BenchmarkId::new("vec_value::iter", BIG_SAMPLE_SIZE),
        &vec,
        |b, vec| {
            b.iter(|| {
                for i in vec.iter() {
                    black_box(*i);
                }
            });
        },
    );
}

fn bench_bucket_vec_iter_rev(c: &mut Criterion) {
    let vec = (0..BIG_SAMPLE_SIZE)
        .into_iter()
        .map(|value| value as i32)
        .collect::<BucketVec<i32, QuadraticConfig>>();
    c.bench_with_input(
        BenchmarkId::new("bucket_vec::iter.rev()", BIG_SAMPLE_SIZE),
        &vec,
        |b, vec| {
            b.iter(|| {
                for i in vec.iter().rev() {
                    black_box(*i);
                }
            });
        },
    );
}

fn bench_vec_box_iter_rev(c: &mut Criterion) {
    let vec = (0..BIG_SAMPLE_SIZE)
        .into_iter()
        .map(|value| Box::new(value as i32))
        .collect::<Vec<Box<i32>>>();
    c.bench_with_input(
        BenchmarkId::new("vec_box::iter.rev()", BIG_SAMPLE_SIZE),
        &vec,
        |b, vec| {
            b.iter(|| {
                for i in vec.iter().rev() {
                    black_box(**i);
                }
            });
        },
    );
}

fn bench_vec_value_iter_rev(c: &mut Criterion) {
    let vec = vec![0; BIG_SAMPLE_SIZE];
    c.bench_with_input(
        BenchmarkId::new("vec_value::iter.rev()", BIG_SAMPLE_SIZE),
        &vec,
        |b, vec| {
            b.iter(|| {
                for i in vec.iter().rev() {
                    black_box(*i);
                }
            });
        },
    );
}

fn bench_bucket_vec_iter_mut(c: &mut Criterion) {
    let vec = vec![0; BIG_SAMPLE_SIZE]
        .into_iter()
        .collect::<BucketVec<i32, QuadraticConfig>>();
    c.bench_with_input(
        BenchmarkId::new("bucket_vec::iter_mut", BIG_SAMPLE_SIZE),
        &BIG_SAMPLE_SIZE,
        move |b, _size| {
            b.iter_batched_ref(
                || vec.clone(),
                |vec| {
                    for i in vec.iter_mut() {
                        black_box(*i);
                    }
                },
                BatchSize::SmallInput,
            );
        },
    );
}

fn bench_vec_box_iter_mut(c: &mut Criterion) {
    let vec = (0..BIG_SAMPLE_SIZE)
        .into_iter()
        .map(|value| Box::new(value as i32))
        .collect::<Vec<Box<i32>>>();
    c.bench_with_input(
        BenchmarkId::new("vec_box::iter_mut", BIG_SAMPLE_SIZE),
        &BIG_SAMPLE_SIZE,
        move |b, _size| {
            b.iter_batched_ref(
                || vec.clone(),
                |vec| {
                    for i in vec.iter_mut() {
                        black_box(**i);
                    }
                },
                BatchSize::SmallInput,
            );
        },
    );
}

fn bench_vec_value_iter_mut(c: &mut Criterion) {
    let vec = vec![0; BIG_SAMPLE_SIZE];
    c.bench_with_input(
        BenchmarkId::new("vec_value::iter_mut", BIG_SAMPLE_SIZE),
        &BIG_SAMPLE_SIZE,
        move |b, _size| {
            b.iter_batched_ref(
                || vec.clone(),
                |vec| {
                    for i in vec.iter_mut() {
                        black_box(*i);
                    }
                },
                BatchSize::SmallInput,
            );
        },
    );
}

criterion_group!(
    bench_push,
    bench_bucket_vec_push,
    bench_bucket_vec_push_get,
    bench_vec_box_push,
    bench_vec_value_push
);
criterion_group!(
    bench_get,
    bench_bucket_vec_get_fast_config,
    bench_bucket_vec_get_medium_config,
    bench_bucket_vec_get_slow_config,
    bench_vec_box_get,
    bench_vec_value_get,
);
criterion_group!(
    bench_iter,
    bench_bucket_vec_iter,
    bench_vec_box_iter,
    bench_vec_value_iter,
);
criterion_group!(
    bench_iter_rev,
    bench_bucket_vec_iter_rev,
    bench_vec_box_iter_rev,
    bench_vec_value_iter_rev,
);
criterion_group!(
    bench_iter_mut,
    bench_bucket_vec_iter_mut,
    bench_vec_box_iter_mut,
    bench_vec_value_iter_mut,
);
criterion_main!(
    bench_push,
    bench_get,
    bench_iter,
    bench_iter_rev,
    bench_iter_mut,
);
